File: android/app/src/main/java/com/blokqr/app/crypto/CapabilityUrl.kt
package com.blokqr.app.crypto

import android.net.Uri
import kotlin.math.ln

/**
 * Détection côté client des « capability-URL » (liens personnels à usage unique).
 *
 * Avant de demander l'analyse PROFONDE (qui transmet l'URL complète à la
 * passerelle), l'application détecte les URL porteuses d'un jeton personnel
 * (réinitialisation, désinscription, lien magique, JWT...). Elle peut alors
 * demander le consentement explicite de l'utilisateur plutôt que d'exposer ce
 * secret. La même logique tourne côté serveur (défense en profondeur).
 */
object CapabilityUrl {

    private val SENSITIVE = Regex(
        "(token|reset|verify|confirm|magic|invite|unsubscribe|activation|" +
            "auth|session|sso|otp|signature|sig|key|secret|access)",
        RegexOption.IGNORE_CASE,
    )
    private val JWT = Regex("\\beyJ[A-Za-z0-9_-]{10,}\\.[A-Za-z0-9_-]{10,}\\.[A-Za-z0-9_-]{6,}\\b")

    data class Assessment(val isCapability: Boolean, val reason: String = "")

    fun assess(url: String): Assessment {
        if (JWT.containsMatchIn(url)) return Assessment(true, "jeton JWT détecté")
        val uri = try { Uri.parse(url) } catch (e: Exception) { return Assessment(false) }

        for (name in uri.queryParameterNames) {
            if (SENSITIVE.containsMatchIn(name)) return Assessment(true, "paramètre « $name »")
            val v = uri.getQueryParameter(name) ?: ""
            if (looksRandom(v)) return Assessment(true, "valeur à forte entropie")
        }
        val path = uri.path ?: ""
        for (seg in path.split("/")) {
            if (looksRandom(seg)) return Assessment(true, "segment de chemin aléatoire")
        }
        if (SENSITIVE.containsMatchIn(path)) return Assessment(true, "chemin sensible")
        return Assessment(false)
    }

    private fun looksRandom(s: String): Boolean = s.length >= 20 && shannon(s) >= 3.5

    private fun shannon(s: String): Double {
        if (s.isEmpty()) return 0.0
        val counts = s.groupingBy { it }.eachCount()
        val n = s.length.toDouble()
        return counts.values.sumOf {
            val p = it / n
            -p * (ln(p) / ln(2.0))
        }
    }
}

File: backend/app/analyzers/capability_url.py
"""
Détection des « capability-URL » (URL porteuses de capacité/jeton).

Faille identifiée : même via OHTTP (IP masquée), envoyer l'URL complète au
palier profond expose l'URL à la passerelle. Or certaines URL SONT un
identifiant : lien de réinitialisation de mot de passe, de désinscription
(contenant l'e-mail), jeton de session, lien magique. Les analyser en profondeur
revient à divulguer un secret.

Ce module détecte ces motifs côté serveur (et la même logique tourne côté
client) afin de NE PAS lancer l'analyse profonde sans consentement explicite, et
de signaler le risque de confidentialité à l'utilisateur.
"""
from __future__ import annotations

import math
import re
from dataclasses import dataclass
from urllib.parse import urlsplit, parse_qsl

from app.schemas import Severity, Signal

_SOURCE = "capability"

# Mots-clés de chemin/paramètre typiques d'un lien à usage unique.
_SENSITIVE_KEYS = re.compile(
    r"(token|reset|verify|confirm|magic|invite|unsubscribe|activation|"
    r"auth|session|sso|otp|signature|sig|key|secret|access)",
    re.IGNORECASE,
)
_JWT = re.compile(r"\beyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{6,}\b")


def _shannon_entropy(s: str) -> float:
    if not s:
        return 0.0
    counts = {c: s.count(c) for c in set(s)}
    n = len(s)
    return -sum((c / n) * math.log2(c / n) for c in counts.values())


def _looks_random(token: str) -> bool:
    """Segment long et à forte entropie => probable jeton."""
    return len(token) >= 20 and _shannon_entropy(token) >= 3.5


@dataclass
class CapabilityVerdict:
    is_capability: bool
    reason: str = ""


def assess_capability(url: str) -> CapabilityVerdict:
    parts = urlsplit(url)

    if _JWT.search(url):
        return CapabilityVerdict(True, "jeton JWT détecté dans l'URL")

    # Paramètres de requête sensibles ou à forte entropie.
    for key, value in parse_qsl(parts.query):
        if _SENSITIVE_KEYS.search(key):
            return CapabilityVerdict(True, f"paramètre sensible « {key} »")
        if _looks_random(value):
            return CapabilityVerdict(True, f"valeur à forte entropie (« {key} »)")

    # Segments de chemin longs et aléatoires.
    for seg in parts.path.split("/"):
        if _looks_random(seg):
            return CapabilityVerdict(True, "segment de chemin à forte entropie")
    if _SENSITIVE_KEYS.search(parts.path):
        return CapabilityVerdict(True, "chemin sensible (réinitialisation/confirmation)")

    return CapabilityVerdict(False)


def capability_signal(url: str) -> Signal | None:
    """Signal informatif (n'augmente pas le score, protège la vie privée)."""
    v = assess_capability(url)
    if not v.is_capability:
        return None
    return Signal(
        code="capability_url",
        title="Lien personnel à usage unique",
        detail=(
            f"Cette URL semble contenir un jeton personnel ({v.reason}). "
            "L'analyse approfondie est suspendue par défaut pour ne pas exposer "
            "ce secret ; relancez-la explicitement si vous faites confiance à la source."
        ),
        severity=Severity.INFO,
        weight=0,
        source=_SOURCE,
    )
